docs/sphinx/examples/python/qmcmc.py

```diff
@@ -0,0 +1,184 @@
+        count_problem_terms =  count_problem_terms + 1
+
+    # mixer hamiltonian
+    for j in range(1, nqubits):
+        H_mix += spin.x(j)
+        H_list.append(gamma * spin.x(j))
+
+    # final hamiltonian
+    alpha = np.linalg.norm(H_mix.to_matrix()) / np.linalg.norm(H_prob.to_matrix())
+    for i in range(count_problem_terms):
+        H_list[i] = H_list[i] * alpha
+
+    ordered_H = []
+    for term in H_list:
+        adjTerm = term
+        for i in range(term.get_qubit_count(), nqubits):
+            adjTerm = term * spin.i(i)
+        ordered_H.append(adjTerm)
+
+    return ordered_H
+
+# This is the first-order Trotter gate decomposition
+def trotter_circuit(kernel, qreg, hk, dt, n_qubits):
+    for term in hk:
+        qbitSupport = []
+        pauliString = str(term).split(' ')[1].rstrip()
+        coef = term.get_coefficient()
+        for i in range(n_qubits):
+            pauli = pauliString[i]
+            if (pauli == 'X'):
+                kernel.h(qreg[i])
+                qbitSupport.append(i)
+            elif (pauli == 'Y'):
+                kernel.rx(np.pi/2.0, qreg[i])
+                qbitSupport.append(i)
+            elif (pauli == 'Z'):
+                qbitSupport.append(i)
+
+        # Apply cnots
+        for j in range(len(qbitSupport)-1):
+            kernel.cx(qreg[qbitSupport[j]], qreg[qbitSupport[j+1]])
+
+        if (len(qbitSupport) > 0):
+            kernel.rz(2.0 * dt * coef.real, qreg[qbitSupport[-1]])
+
+        for j in range(len(qbitSupport)-1,0,-1):
+            kernel.cx(qreg[qbitSupport[j-1]], qreg[qbitSupport[j]])
+        for i in range(n_qubits-1, -1, -1):
+
+            pauli = pauliString[i]
+
+            if (pauli == 'X'):
+                kernel.h(qreg[i])
+            elif (pauli == 'Y'):
+                kernel.rx(-np.pi/2.0, qreg[i])
+            elif (pauli == 'Z'):
+                pass
+    return kernel
+
+
+def main():
+
+    random.seed(41)
+
+    # Argument parsing
+    parser = argparse.ArgumentParser(description='Quantum Enhanced Markov Chain Monte Carlo')
+    parser.add_argument('--num_iterations', type=int, default=20, help='Number of iterations')
+    parser.add_argument('--nqubits', type=int, default=10, help='Number of qubits')
+    parser.add_argument('--temperature', type=float, default=0.1, help='Temperature')
+    parser.add_argument('--shots_count', type=int, default=10, help='Number of shots')
+    args = parser.parse_args()
+
+    num_iterations = args.num_iterations
+    nqubits = args.nqubits
+    T = args.temperature
+    shots_count = args.shots_count
+
+    # Specify couplings J's and fields h's.
+    J = np.ones((nqubits,nqubits))
+    h = np.ones(nqubits)
+
+    # Generate a random initial bitstring instead of a spin config
+    # Convert it to its equivalent spin config later when needed
+    s = generate_random_bitstring(nqubits)
+
+    # Create a circuit and allocate qubits
+    kernel = cudaq.make_kernel()
```

Comment:
Is the kernel supposed to grow / build on itself each iteration? I notice in the appendix of the paper you are just interested in computing `|psi> = exp(-iHt)|s>` each iteration, so should this `make_kernel()` and subsequent `qalloc` call be in the loop body right before `initial_state`?

Reply:
Yes, that can be done. The way I am doing things now is by resetting the kernel at the end of each iteration. I am not sure which way is more performant.

Reply:
That reset method on kernel_builder is for resetting the qubit's state.